lxd/storage/drivers/driver_btrfs.go

```diff
@@ -0,0 +1,1514 @@
+	if vol.contentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if vol.volType != VolumeTypeCustom {
+		return fmt.Errorf("Volume type not supported")
+	}
+
+	return d.SetVolumeQuota(vol.volType, vol.name, vol.config["size"], nil)
+}
+
+// RenameVolume renames a volume and its snapshots.
+func (d *btrfs) RenameVolume(volType VolumeType, volName string, newVolName string, op *operations.Operation) error {
+	srcVolumePath := GetVolumeMountPath(d.name, volType, volName)
+	dstVolumePath := GetVolumeMountPath(d.name, volType, newVolName)
+
+	err := os.Rename(srcVolumePath, dstVolumePath)
+	if err != nil {
+		return err
+	}
+
+	srcSnapshotDir := GetVolumeSnapshotDir(d.name, volType, volName)
+	dstSnapshotDir := GetVolumeSnapshotDir(d.name, volType, newVolName)
+
+	if shared.PathExists(srcSnapshotDir) {
+		err = os.Rename(srcSnapshotDir, dstSnapshotDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RestoreVolume restores a volume from a snapshot.
+func (d *btrfs) RestoreVolume(vol Volume, snapshotName string, op *operations.Operation) error {
+	// Create a backup so we can revert.
+	backupSubvolume := fmt.Sprintf("%s.tmp", vol.MountPath())
+
+	err := os.Rename(vol.MountPath(), backupSubvolume)
+	if err != nil {
+		return err
+	}
+
+	undoSnapshot := true
+	defer func() {
+		if undoSnapshot {
+			os.Rename(vol.MountPath(), backupSubvolume)
+		}
+	}()
+
+	source := GetVolumeMountPath(d.name, vol.volType, GetSnapshotVolumeName(vol.name, snapshotName))
+
+	err = btrfsSubvolumesSnapshotCreate(source, vol.MountPath(), false, true, d.state.OS.RunningInUserNS)
+	if err != nil {
+		return err
+	}
+
+	undoSnapshot = false
+
+	// Remove the backup subvolume
+	return btrfsSubvolumesDelete(backupSubvolume)
+}
+
+// DeleteVolume deletes a volume of the storage device. If any snapshots of the volume remain then
+// this function will return an error.
+func (d *btrfs) DeleteVolume(volType VolumeType, volName string, op *operations.Operation) error {
+	snapshots, err := d.VolumeSnapshots(volType, volName, op)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) > 0 {
+		return fmt.Errorf("Cannot remove a volume that has snapshots")
+	}
+
+	volPath := GetVolumeMountPath(d.name, volType, volName)
+
+	if !shared.PathExists(volPath) || !isBtrfsSubvolume(volPath) {
+		return nil
+	}
+
+	err = btrfsSubvolumeDelete(volPath)
+	if err != nil {
+		return err
+	}
+
+	// Although the volume snapshot directory should already be removed, lets remove it here
+	// to just in case the top-level directory is left.
+	err = deleteParentSnapshotDirIfEmpty(d.name, volType, volName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// MountVolume simulates mounting a volume. As dir driver doesn't have volumes to mount it returns
+// false indicating that there is no need to issue an unmount.
+func (d *btrfs) MountVolume(volType VolumeType, volName string, op *operations.Operation) (bool, error) {
+	_, err := d.Mount()
```

Comment:
This is wrong, we never need to call `d.Mount()` as the pool is always mounted when loaded (exception being Create).